compiler/src/typed/typecore.re

```diff
@@ -738,13 +740,23 @@ and type_expect_ =
         | Not_found => None
         };
 
-      switch (get_path(ty_expected)) {
-      | None => (newvar(), None)
-      | op => (ty_expected, op)
+      let expected_opath = get_path(ty_expected);
+      let opt_exp_opath = Option.bind(opt_exp, exp => get_path(exp.exp_type));
+      switch (expected_opath, opt_exp_opath) {
+      | (None, None) => (newvar(), None)
+      | (Some(_), None)
+      | (Some((_, _, true)), Some(_)) => (ty_expected, expected_opath)
+      | (None | Some((_, _, false)), Some((_, p', _))) =>
+        let decl = Env.find_type(p', env);
+        begin_def();
+        let ty = newconstr(p', instance_list(env, decl.type_params));
+        end_def();
+        generalize_structure(ty);
+        (ty, opt_exp_opath);
       };
     };
 
-    let closed = true;
+    let closed = Option.is_none(b);
```

Comment:
Is this right? We don't have open record types at all.

Reply:
Not sure, maybe this isn't necessary. To be honest, most of the details of the typing implementation go over my head and I only changed this because this is how it appears in the OCaml compiler, and so I figured that maybe the new inclusion of record extensions could influence what the "correct" value of this variable should be.